optional_plugins/golang/avocado_golang/__init__.py

```diff
@@ -0,0 +1,225 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2017
+# Authors: Amador Pahim <apahim@redhat.com>
+
+"""
+Plugin to run Golang tests in Avocado
+"""
+
+import fnmatch
+import glob
+import os
+import re
+
+from avocado.core import loader
+from avocado.core import output
+from avocado.core import test
+from avocado.core.plugin_interfaces import CLI
+from avocado.utils import path as utils_path
+from avocado.utils import process
+
+
+_GO_BIN = utils_path.find_command('go')
+
+
+class GolangTest(test.SimpleTest):
+
+    """
+    Run a Golang Test command as a SIMPLE test.
+    """
+
+    def __init__(self, name,
+                 params=None,
+                 base_logdir=None,
+                 job=None,
+                 subtest=None):
+        super(GolangTest, self).__init__(name, params, base_logdir, job)
+        self.subtest = subtest
+
+    @property
+    def filename(self):
+        """
+        Returns the path of the golang test suite.
+        """
+        return self.name.name.split(':')[0]
+
+    def test(self):
+        """
+        Create the Golang command and execute it.
+        """
+        test_name = '%s$' % self.name.name.split(':')[1]
+        if self.subtest is not None:
+            test_name += '/%s' % self.subtest
+
+        cmd = '%s test -v %s -run %s' % (_GO_BIN, self.filename, test_name)
+
+        result = process.run(cmd, ignore_status=True)
+        if result.exit_status != 0:
+            self.fail('Golang Test execution returned a '
+                      'non-0 exit code (%s)' % result)
+
+
+class NotGolangTest(object):
+
+    """
+    Not a golang test (for reporting purposes)
+    """
+
+
+class GolangLoader(loader.TestLoader):
+    """
+    Golang loader class
+    """
+    name = "golang"
+
+    def __init__(self, args, extra_params):
+        super(GolangLoader, self).__init__(args, extra_params)
+
+    def discover(self, url, which_tests=loader.DEFAULT):
+        if url is None:
+            return []
+
+        avocado_suite = []
+        package_paths = []
+        test_files = []
+        subtest = None
+        tests_filter = None
+
+        if ':' in url:
+            url, _tests_filter = url.split(':', 1)
+            parsed_filter = re.split(r'(?<!\\)/', _tests_filter, 1)
+            _tests_filter = parsed_filter[0]
+            if len(parsed_filter) > 1:
+                subtest = parsed_filter[1]
+            tests_filter = re.compile(_tests_filter)
+
+        # When a file is provided
+        if os.path.isfile(url):
```

Comment:
Isn't this treatment unnecessary complex? I'd expect it to work the same way as `GOROOT` so plain `package_paths.append('.')` would IMO work. Anyway I don't know that for sure and this (even so it's long) should work as well. We can modify it based on GO users requests later.

Reply:
As far as I could understand the `go test` behavior, there are some differences when you use a package reference when compared to file path references. Depending on the test and on $PWD, modules might not be loaded properly. That's why I think we should have a separate treatment for packages. If you don't mind, let's start with this and see where it leads us.

Reply:
yep, no problem, it just looked unnecessary complex but I only experienced with it briefly... We'll hear about that form go devels anyway 😄